### PR TITLE
Add assemblies and packages for new code style analyzer distribution

### LINKED ISSUE #18278: Create new assemblies for NuGet-distributed analyzers
We need new assemblies to define analyzers and code fixes in order to distribute them on NuGet. This is a prerequisite to #17182.

### PR DESCRIPTION
**Pull request by sharwell**
Fixes #18278.

New analyzer packages:

* **Microsoft.CodeAnalysis.CSharp.CodeStyle**
* **Microsoft.CodeAnalysis.VisualBasic.CodeStyle**

New projects containing code distributed with the above packages:

* **Microsoft.CodeAnalysis.CodeStyle** (shared code for analyzers, no workspace reference)
* **Microsoft.CodeAnalysis.CodeStyle.Fixes** (shared code for fixers)
* **Microsoft.CodeAnalysis.CSharp.CodeStyle** (C# code style analyzers, no workspace reference)
* **Microsoft.CodeAnalysis.CSharp.CodeStyle.Fixes** (C# code style fixes)
* **Microsoft.CodeAnalysis.VisualBasic.CodeStyle** (VB code style analyzers, no workspace reference)
* **Microsoft.CodeAnalysis.VisualBasic.CodeStyle.Fixes** (VB code style fixes)

New test projects:

* **Microsoft.CodeAnalysis.CodeStyle.UnitTests**
* **Microsoft.CodeAnalysis.CodeStyle.CSharp.UnitTests**
* **Microsoft.CodeAnalysis.CodeStyle.VisualBasic.UnitTests**

#### Notes for specific reviewers:

@srivatsn: Please review the inclusion of **install.ps1** and **uninstall.ps1** in the new analyzer NuGet packages. Are these still necessary (or does NuGet automatically add references now), and if so, is this how you would include them?

@jaredpar I know you care about infrastructure, and also thanks for all the help leading up to this! 👏 

@dpoeschl Is this enough of a foundation for us to work in parallel on the migration?

@Pilchie Please review the proposed NuGet package names and manner of distribution, etc.

@tmat Please review the change to **BuildNuGets.csx**.

### CODE REVIEW

File: src/NuGet/Microsoft.CodeAnalysis.CSharp.CodeStyle.nuspec [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    <file src="Dlls\CSharpCodeStyleFixes\Microsoft.CodeAnalysis.CSharp.CodeStyle.Fixes.dll" target="analyzers\dotnet\cs" />
+    <file src="Dlls\CSharpCodeStyleFixes\Microsoft.CodeAnalysis.CSharp.CodeStyle.Fixes.xml" target="analyzers\dotnet\cs" />
+    <file src="Dlls\CodeStyle\Microsoft.CodeAnalysis.CodeStyle.dll" target="analyzers\dotnet\cs" />
+    <file src="Dlls\CodeStyle\Microsoft.CodeAnalysis.CodeStyle.xml" target="analyzers\dotnet\cs" />
+    <file src="Dlls\CodeStyleFixes\Microsoft.CodeAnalysis.CodeStyle.Fixes.dll" target="analyzers\dotnet\cs" />
+    <file src="Dlls\CodeStyleFixes\Microsoft.CodeAnalysis.CodeStyle.Fixes.xml" target="analyzers\dotnet\cs" />
+    <file src="$thirdPartyNoticesPath$" target="" />
+
+    <!-- Scripts -->
+    <file src="..\..\src\Setup\PowerShell\install.ps1" target="tools\" />
```
Comment (srivatsn) [inline src/NuGet/Microsoft.CodeAnalysis.CSharp.CodeStyle.nuspec]: Yes you need these if someone's installing the package to a packages.config-based project which is still the majority of projects out there.

File: src/NuGet/BuildNuGets.csx [hunk 2]
```diff
@@ -122,6 +124,8 @@ string[] TestPackageNames = {
 var PreReleaseOnlyPackages = new HashSet<string>
 {
     "Microsoft.CodeAnalysis.Build.Tasks",
+    "Microsoft.CodeAnalysis.CSharp.CodeStyle",
+    "Microsoft.CodeAnalysis.VisualBasic.CodeStyle",
```
Comment (Pilchie) [inline src/NuGet/BuildNuGets.csx]: why are these pre-release only?  When we ship a final release, we'll want to upload these?

File: src/NuGet/Microsoft.CodeAnalysis.CSharp.CodeStyle.nuspec [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    <licenseUrl>$licenseUrl$</licenseUrl>
+    <projectUrl>$projectUrl$</projectUrl>
+    <releaseNotes>$releaseNotes$</releaseNotes>
+    <tags>$tags$</tags>
+    <serviceable>true</serviceable>
+    <developmentDependency>true</developmentDependency>
+  </metadata>
+  <files>
+    <file src="Dlls\CSharpCodeStyle\Microsoft.CodeAnalysis.CSharp.CodeStyle.dll" target="analyzers\dotnet\cs" />
+    <file src="Dlls\CSharpCodeStyle\Microsoft.CodeAnalysis.CSharp.CodeStyle.xml" target="analyzers\dotnet\cs" />
```
Comment (Pilchie) [inline src/NuGet/Microsoft.CodeAnalysis.CSharp.CodeStyle.nuspec]: Not sure if it makes sense to distribute these for analyzers.

File: src/NuGet/BuildNuGets.csx (hunk 2 above)
Comment (sharwell) [inline src/NuGet/BuildNuGets.csx]: I literally took a guess and went with it, and then added @tmat as a required reviewer. 😆

File: src/NuGet/Microsoft.CodeAnalysis.CSharp.CodeStyle.nuspec (hunk 3 above)
Comment (sharwell) [inline src/NuGet/Microsoft.CodeAnalysis.CSharp.CodeStyle.nuspec]: I was thinking the same thing. I'm pretty sure the answer is we don't need them, so I'll go ahead and remove them now.

**Comment by sharwell** [comment]
@Pilchie 

> My big concern is about how we do this transition with incurring perf/dll load issues in the default VS cases.

Oh, I'm worried about this too, but the big perf hit isn't going to be the DLL load. The switch from IDE- to NuGet-based analyzers will cause the code style analyzers to be regularly run against entire code bases, as opposed to only running them against files open in the IDE. I would not be surprised to find this leads to required perf work in both the code style analyzers themselves and the analyzer driver. On the bright side, improvements to the latter could lead to a big perf win for external analyzer projects which are already being shipped as NuGet packages.

**Comment by jaredpar** [comment]
retest ubuntu_14_debug_prtest please

File: src/NuGet/BuildNuGets.csx (hunk 2 above)
Comment (tmat) [inline src/NuGet/BuildNuGets.csx]: As the comment above says, are we not going to publish these on nuget?

---
In reply to: [108894557](https://github.com/dotnet/roslyn/pull/18285#discussion_r108894557) [](ancestors = 108894557)

Comment (sharwell) [inline src/NuGet/BuildNuGets.csx]: We are planning to publish these on NuGet. Does that mean I should remove them from `PreReleaseOnlyPackages` from the start (i.e. before they would actually be considered "ready for public release")?

Comment (tmat) [inline src/NuGet/BuildNuGets.csx]: Yes 

CodeFlow.

---
In reply to: [109054195](https://github.com/dotnet/roslyn/pull/18285#discussion_r109054195) [](ancestors = 109054195)

**Comment by alrz** [comment]
@sharwell Is there any plan to move fxcop analyzers to these projects, e.g. https://github.com/dotnet/roslyn/issues/16621 

https://github.com/dotnet/roslyn-analyzers/blob/569bd373a4831d3035597197e02980b57602a7f2/src/Microsoft.QualityGuidelines.Analyzers/Core/MarkMembersAsStatic.cs

**Comment by sharwell** [comment]
@alrz It was not my original intent for the FxCop analyzers to move to these packages. The main short-term goal is to move several of the existing IDExxxx analyzers away from being IDE-specific to being able to run during command line builds (and thus on CI servers).

We're facing some big hurdles right now related to the fact that some of these analyzers use APIs that are only available while working in the IDE (mostly related to configurable options). I wanted to get the first step of creating these projects out of the way so the 50 changed files you see here don't interfere with our ability to review upcoming proposals related to exposing options to analyzer authors in a uniform way.

**Comment by ErikSchierboom** [comment]
@sharwell I have two questions, which you hopefully can answer:

1. Are these new packages published on NuGet yet? I cannot seem to [find them](https://www.nuget.org/packages?q=Microsoft.CodeAnalysis.CSharp.CodeStyle).
1. Will these new packages work with .NET Core projects?

**Comment by jmarolf** [comment]
@ErikSchierboom 
1. they have not been published yet, but they should go on https://dotnet.myget.org/gallery/roslyn eventually
2. They are netstandard1.3 so yes

**Comment by ErikSchierboom** [comment]
@jmarolf Okay, thanks for the response! Looking forward to them being released.

**Pull request merged**